### PR TITLE
New module contract in v9.0

### PR DESCRIPTION
**Pull request by carlosdauden**
This PR renames and converts to new API the module _account_analytic_analysis_recurring_ from v7, which is itself an extraction of the recurring invoicing feature from Odoo's v8 _account_analytic_analysis_ module, and integrates the functionality of the following modules:
- account_analytic_analysis_recurring.
- contract_journal
- contract_discount
- contract_recurring_invoicing_marker
- contract_show_recurring_invoice

### CODE REVIEW

File: contract/README.rst [hunk 1]
```diff
@@ -0,0 +1,69 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+    :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
+    :alt: License: AGPL-3
+
+========
+Contract
```
Comment (elicoidal) [inline contract/README.rst]: Contracts for recurrent invoicing

File: contract/README.rst [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+========
+
+ * This module recover contracts management with recurring invoicing functions.
```
Comment (elicoidal) [inline contract/README.rst]: This module forward-port to v9 the contracts management with recurring invoicing functions.

File: contract/README.rst [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+#. A cron is created with daily interval, but if you are in debug mode can
+   click on *Create invoices* to force this action.
+
+
+
+.. image:: https://odoo-community.org/website/image/ir.attachment/5784_f2813bd/datas
+   :alt: Try me on Runbot
+   :target: https://runbot.odoo-community.org/runbot/110/9.0
+
+For further information, please visit:
```
Comment (elicoidal) [inline contract/README.rst]: remove

File: contract/__openerp__.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
+# -*- coding: utf-8 -*-
+# © 2016 Incaser Informatica S.L. - Carlos Dauden
```
Comment (elicoidal) [inline contract/__openerp__.py]: Should you add OpenERP SA in the copyright holder?

File: contract/__openerp__.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+# License AGPL-3 - See http://www.gnu.org/licenses/agpl-3.0.html
+
+{
+    'name': 'Contracts Management recurring',
+    'version': '9.0.1.0.0',
+    'category': 'Other',
+    'license': 'AGPL-3',
+    'author': "OpenERP SA,Odoo Community Association (OCA)",
```
Comment (elicoidal) [inline contract/__openerp__.py]: add Incaser in the list of authors

File: contract/models/contract.py [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+        string='Generate recurring invoices automatically')
+    recurring_rule_type = fields.Selection(
+        [('daily', 'Day(s)'),
+         ('weekly', 'Week(s)'),
+         ('monthly', 'Month(s)'),
+         ('yearly', 'Year(s)'),
+         ],
+        default='monthly',
+        string='Recurrency',
+        help="Invoice automatically repeat at specified interval")
```
Comment (elicoidal) [inline contract/models/contract.py]: Specify Interval for automatic invoice generation.

**Comment by elicoidal** [comment]
some details. Other than that LGTM (no test)

File: contract/models/contract.py [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+        default=fields.Date.context_today,
+        string='Date of Next Invoice')
+    journal_id = fields.Many2one(
+        'account.journal',
+        string='Journal',
+        default=_default_journal,
+        domain="[('type', '=', 'sale'),('company_id', '=', company_id)]")
+
+    @api.multi
+    def copy(self, default=None):
```
Comment (pedrobaeza) [inline contract/models/contract.py]: This is better to be put as `copy=False` in the field `recurring_next_date`

File: contract/models/contract.py [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+        'product.uom', string='Unit of Measure', required=True)
+    price_unit = fields.Float('Unit Price', required=True)
+    price_subtotal = fields.Float(
+        compute='_compute_price_subtotal',
+        digits_compute=dp.get_precision('Account'),
+        string='Sub Total')
+    discount = fields.Float(
+        string='Discount (%)',
+        digits=dp.get_precision('Discount'),
+        copy=True,
```
Comment (pedrobaeza) [inline contract/models/contract.py]: This is redundant

File: contract/models/contract.py [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+    @api.model
+    def _insert_markers(self, line, date_start, next_date, date_format):
+        line = line.replace('#START#', date_start.strftime(date_format))
+        date_end = next_date - relativedelta(days=1)
+        line = line.replace('#END#', date_end.strftime(date_format))
+        return line
+
+    @api.model
+    def _prepare_invoice_line(self, line, invoice_id):
+        product = line.product_id
```
Comment (pedrobaeza) [inline contract/models/contract.py]: Here you have to use the same technique as in https://github.com/OCA/commission/pull/67/files#diff-73f037396680f271cd1a6b014c9935ecR86 to use the invoice line `onchange_product_id`and this way, you won't need to make a lot of glue modules that inject their values on it.

File: contract/models/contract.py [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+        fpos = partner.property_account_position_id
+        journal = contract.journal_id or self.env['account.journal'].search(
+            [('type', '=', 'sale'),
+             ('company_id', '=', contract.company_id.id)],
+            limit=1)
+        if not journal:
+            raise ValidationError(
+                _("Please define a sale journal for the company '%s'.") %
+                (contract.company_id.name or '',))
+        inv_data = {
```
Comment (pedrobaeza) [inline contract/models/contract.py]: The same for `onchange_partner_id`.

File: contract/models/contract.py [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+                new_date = next_date + relativedelta(months=interval)
+            ctx = self.env.context.copy()
+            ctx.update({
+                'old_date': old_date,
+                'next_date': new_date,
+                # Force company for correct evaluate domain access rules
+                'force_company': contract.company_id.id,
+            })
+            # Re-read contract with correct company
+            contract = self.with_context(ctx).browse(contract.id)
```
Comment (pedrobaeza) [inline contract/models/contract.py]: Simply make `contract = contract.with_context(ctx)`

File: contract/models/contract.py [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+        current_date = time.strftime('%Y-%m-%d')
+        contracts = self.search(
+            [('recurring_next_date', '<=', current_date),
+             ('account_type', '=', 'normal'),
+             ('recurring_invoices', '=', True)])
+        for contract in contracts:
+            next_date = fields.Date.from_string(
+                contract.recurring_next_date or fields.Date.today())
+            interval = contract.recurring_interval
+            old_date = next_date
```
Comment (pedrobaeza) [inline contract/models/contract.py]: Why not using directly `old_date` in all expressions and remove `next_date`?

**Comment by pedrobaeza** [comment]
I think you have forgotten to integrate this view: https://github.com/OCA/contract/blob/8.0/contract_recurring_invoicing_marker/views/account_analytic_account_view.xml

File: contract/tests/test_contract.py [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+                         self.contract.date_start)
+
+    def test_check_journal(self):
+        contract_no_journal = self.contract.copy()
+        contract_no_journal.journal_id = False
+        journal = self.env['account.journal'].search([('type', '=', 'sale')])
+        journal.write({'type': 'general'})
+        with self.assertRaises(ValidationError):
+            contract_no_journal.recurring_create_invoice()
+        journal.write({'type': 'sale'})
```
Comment (pedrobaeza) [inline contract/tests/test_contract.py]: You don't need to write again the previous type, as all operations are rolled back when the method finishes, so no need to manually restore it.

File: contract/README.rst [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+
+To use this module, you need to:
+
+#. Go to Sales -> Contracts and select or create a new contract.
+#. Check *Generate recurring invoices automatically*.
+#. Fill fields and add new lines.
+#. To view discount field set *Discount on lines* in user access rights.
+#. A cron is created with daily interval, but if you are in debug mode can
+   click on *Create invoices* to force this action.
+
```
Comment (pedrobaeza) [inline contract/README.rst]: Mention here about the possibility of writing the markers, and the button for showing the created invoices.

**Comment by carlosdauden** [comment]
Changes done

File: contract/README.rst [hunk 15]
```diff
@@ -0,0 +1,67 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+    :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
+    :alt: License: AGPL-3
+
+=================================
+Contracts for recurrent invoicing
+=================================
+
+ * This module forward-port to v9 the contracts management with recurring
```
Comment (pedrobaeza) [inline contract/README.rst]: Don't put here a `*`

File: contract/README.rst [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+
+Usage
+=====
+
+To use this module, you need to:
+
+#. Go to Sales -> Contracts and select or create a new contract.
+#. Check *Generate recurring invoices automatically*.
+#. Fill fields and add new lines. You have the possibility to use markers in
+   in description field to show init date and end of invoiced period.
```
Comment (pedrobaeza) [inline contract/README.rst]: s/in/the

Comment (pedrobaeza) [inline contract/README.rst]: s/to show init date and end of invoiced period/to show the start and end date of the invoiced period

File: contract/README.rst [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+#. To view discount field set *Discount on lines* in user access rights.
```
Comment (pedrobaeza) [inline contract/README.rst]: This goes into configuration section

File: contract/README.rst [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+#. A cron is created with daily interval, but if you are in debug mode can
+   click on *Create invoices* to force this action.
+#. Click *Show invoices* link to show all invoices created by the contract.
```
Comment (pedrobaeza) [inline contract/README.rst]: Rename this to "Show recurring invoices"

File: contract/README.rst [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+#. Press *Invoices* button to show all invoices related with the contract.
```
Comment (pedrobaeza) [inline contract/README.rst]: Rename this to "Show invoices"

File: contract/i18n/es.po [hunk 20]
```diff
@@ -0,0 +1,302 @@
...
+msgid "account.analytic.invoice.line"
+msgstr "account.analytic.invoice.line"
+
+#. module: contract
+#: model:ir.ui.view,arch_db:contract.account_analytic_account_recurring_form_form
+msgid "⇒ Show invoices"
+msgstr "⇒ Mostrar facturas"
+
+#, fuzzy
+#~ msgid "Recurring Invoice"
```
Comment (pedrobaeza) [inline contract/i18n/es.po]: Why is this commented? Remove them

File: contract/models/contract.py [hunk 21]
```diff
@@ -0,0 +1,256 @@
+# -*- coding: utf-8 -*-
+# © 2016 Incaser Informatica S.L. - Carlos Dauden
```
Comment (pedrobaeza) [inline contract/models/contract.py]: Put here the rest of the credits

File: contract/README.rst (hunk 19 above)
Comment (pedrobaeza) [inline contract/README.rst]: This shouldn't go into this module, because it's an independent feature. Keep it on the original module.

**Comment by carlosdauden** [comment]
Changes done

**Comment by pedrobaeza** [comment]
:+1: 

@rafaelbn, @sergio-incaser please review

**Comment by sergio-teruel** [comment]
:+1:

**Comment by guewen** [comment]
:+1:

**Comment by pedrobaeza** [comment]
Don't merge this yet, please.

**Comment by sergio-teruel** [comment]
:+1:

**Pull request merged**